Source/Core/Common/WorkQueueThread.h

```diff
@@ -20,89 +22,148 @@ class WorkQueueThread
 {
 public:
   WorkQueueThread() = default;
-  WorkQueueThread(std::function<void(T)> function) { Reset(std::move(function)); }
+  WorkQueueThread(const std::string_view name, std::function<void(T)> function)
+  {
+    Reset(name, std::move(function));
+  }
   ~WorkQueueThread() { Shutdown(); }
-  void Reset(std::function<void(T)> function)
+
+  // Shuts the current work thread down (if any) and starts a new thread with the given function
+  // Note: Some consumers of this API push items to the queue before starting the thread.
+  void Reset(const std::string_view name, std::function<void(T)> function)
   {
     Shutdown();
-    m_shutdown.Clear();
-    m_cancelled.Clear();
+    std::lock_guard lg(m_lock);
+    m_thread_name = name;
+    m_shutdown = false;
     m_function = std::move(function);
     m_thread = std::thread(&WorkQueueThread::ThreadLoop, this);
   }
 
+  // Adds an item to the work queue
   template <typename... Args>
   void EmplaceItem(Args&&... args)
   {
-    if (!m_cancelled.IsSet())
-    {
-      std::lock_guard lg(m_lock);
-      m_items.emplace(std::forward<Args>(args)...);
-    }
-    m_wakeup.Set();
+    std::lock_guard lg(m_lock);
+    if (m_shutdown)
+      return;
+
+    m_items.emplace(std::forward<Args>(args)...);
+    m_idle = false;
+    m_worker_cond_var.notify_one();
   }
 
-  void Clear()
+  // Adds an item to the work queue
+  void Push(T&& item)
   {
-    {
-      std::lock_guard lg(m_lock);
-      m_items = std::queue<T>();
-    }
-    m_wakeup.Set();
+    std::lock_guard lg(m_lock);
+    if (m_shutdown)
+      return;
+
+    m_items.push(std::move(item));
+    m_idle = false;
+    m_worker_cond_var.notify_one();
   }
 
-  void Cancel()
+  // Adds an item to the work queue
+  void Push(const T& item)
   {
-    m_cancelled.Set();
-    Clear();
-    Shutdown();
+    std::lock_guard lg(m_lock);
+    if (m_shutdown)
+      return;
+
+    m_items.push(item);
+    m_idle = false;
+    m_worker_cond_var.notify_one();
   }
 
-  bool IsCancelled() const { return m_cancelled.IsSet(); }
+  // Empties the queue
+  // If the worker polls IsCanceling(), it can abort it's work when Cancelling
+  void Cancel()
+  {
+    std::unique_lock lg(m_lock);
+    if (m_shutdown)
+      return;
 
-  void Shutdown()
+    m_cancelling = true;
```

Comment:
Not sure I follow the cancel logic.  Cancelling empties the queue where the worker thread is in a while loop.  Presumably the thread is waiting in the `if (m_items.empty())` check where the condition variable is.  The `notify_one()` will notify the thread and `m_cancelling` will be set, triggering the wait to end.  But the next iteration of the loop it will just re-enter the `if (m_items.empty())` check and wait on the condition variable again?

Reply:
Yes, we want the loop to run at least once, just so it can reset `m_cancelling` back to false.

This is so `WaitForCompletion` will wait for the cancel.... Though I just noticed we don't set `m_idle` here so there is a potential race condition.

Reply:
I've adjusted `WaitForCompletion` to check `m_cancelling`. It's probably not 100% necessary, but I thought it would be best for sanity.  

Also fixed a potential deadlock with WaitForCompletion and shutdown.